File: benchmarks/families/cnn.py
#!/usr/bin/env python3
"""The CNN family in PyTorch, defined once, driven four ways.

PLAN.md; the counterpart of cnn.cpp and deliberately its mirror image. Same
modes, same positional arguments, same `key=value` output, so run.py drives
either engine by swapping the command prefix.

  cnn.py train    <train_dir> <test_dir> [epochs] [batch,...] [size] [dev] [prec]
  cnn.py infer    <test_dir>             [reps]   [batch,...] [size] [dev] [prec]
  cnn.py capacity <train_dir>            [batch]              [size] [dev] [prec]
  cnn.py quality  <train_dir> <test_dir> [epochs] [batch]     [size] [dev] [prec]

`torchvision.models.resnet50` *is* ResNet-50 v1.5 -- the stride sits in the
3x3 convolution -- which is what cnn.cpp builds from bottleneck blocks
[3,4,6,3] over widths [64,128,256,512]. Using the library definition rather
than hand-rolling one keeps the comparison against the citable network instead
of against our transcription of it.

Images are lazy-loaded per batch from class folders by a DataLoader with
worker processes that decode the JPEGs every epoch, against OpenNN's
per-batch reads of its pre-decoded image cache: at 50,000 x 224x224x3 the
split cannot be resident, so this measures convolution throughput *plus*
input-pipeline efficiency in both engines. The two pipelines are not the same
work -- PT_INPUT=cache feeds PyTorch from OpenNN's cache file to measure how
much of the training margin is the decode.
"""

from __future__ import annotations

import contextlib
import os
import re
import sys
import time
from pathlib import Path

import torch
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import ImageFolder
from torchvision.models import resnet50

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

SEED = 42
WORKERS = int(os.environ.get("PT_WORKERS", "8"))
INPUT = os.environ.get("PT_INPUT", "jpeg")
# Inference in bf16 can keep autocast on (weights re-cast on every call, which
# torch.compile does not fold without freezing) or store the weights in bf16
# once, the way OpenNN keeps a bf16 mirror of its parameters. PT_INFER_CAST
# selects it; the published cell uses whichever measured faster.
INFER_CAST = os.environ.get("PT_INFER_CAST", "autocast")

def report_blas() -> None:
    """Which BLAS this engine dispatches to, printed like OpenNN prints it.

    PyTorch's is fixed at build time -- the wheels are MKL-backed -- while
    OpenNN's is chosen at runtime. Recording both is what stops a comparison
    quietly turning into Eigen against MKL.
    """
    settings = torch.__config__.show()
    match = re.search(r"BLAS_INFO=(\w+)", settings)

    print(f"blas={match.group(1) if match else 'unknown'}")


def build(opts: dict) -> torch.nn.Module:
    """The CNN family. Nothing else here constructs the network."""
    torch.manual_seed(SEED)

    model = resnet50(weights=None, num_classes=1000).to(opts["device"])
    if opts["device"] == "cuda":
        # channels_last is how a convolution reaches the tensor cores; without
        # it PyTorch is measured on a layout it would never ship.
        model = model.to(memory_format=torch.channels_last)
    return model

def resident_mib() -> float:
    """Resident set, MiB. The framework baseline is subtracted from the peak,
    because torch's import alone is ~816 MiB here against OpenNN's 209 -- so a
    raw RSS comparison measures which framework is bigger, not which run costs
    more, and the answer flips with dataset size."""
    try:
        with open("/proc/self/statm") as handle:
            return int(handle.read().split()[1]) * os.sysconf("SC_PAGE_SIZE") / (1024.0 * 1024.0)
    except Exception:
        return 0.0

def parse_opts(argv: list[str], first: int) -> dict:
    def at(index: int, default: str) -> str:
        return argv[index] if len(argv) > index else default

    size = int(at(first, "224"))
    device = at(first + 1, "cuda")
    precision = at(first + 2, "fp32")

    if device == "cuda" and not torch.cuda.is_available():
        raise SystemExit("cuda requested but not available")

    # The runner pins CPU work to performance cores and sets a matched thread
    # count for both engines; honour it, or this side quietly takes every
    # logical CPU and the comparison becomes a thread-count comparison.
    threads = os.environ.get("TORCH_NUM_THREADS") or os.environ.get("OMP_NUM_THREADS")
    if threads:
        torch.set_num_threads(int(threads))

    allow_tf32 = precision != "strict"
    torch.backends.cuda.matmul.allow_tf32 = allow_tf32
    torch.backends.cudnn.allow_tf32 = allow_tf32
    torch.backends.cudnn.benchmark = True

    return {"size": size, "device": device, "precision": precision,
            "autocast": precision == "bf16", "tf32": allow_tf32}

class Folders(ImageFolder):
    """ImageFolder that ignores dot-directories.

    OpenNN's ImageDataset writes its per-batch binary cache into `.cache/`
    beside the class folders, and ImageFolder would otherwise enrol it as a
    1001st class and then fail on it for holding no images. Skipping hidden
    directories keeps both engines reading the same 1000 classes from the same
    tree, which is the point of the shared layout.
    """

    def find_classes(self, directory):
        names = sorted(e.name for e in os.scandir(directory)
                       if e.is_dir() and not e.name.startswith("."))
        return names, {name: index for index, name in enumerate(names)}

class CachedImages(torch.utils.data.Dataset):
    """OpenNN's pre-decoded image cache, read the way OpenNN reads it.

    A controlled variant (`PT_INPUT=cache`), not the published cell. The
    published training cell has PyTorch decode JPEGs per epoch while OpenNN
    reads `.cache/images.bin` -- uint8 HxWxC per image, sorted-folder /
    sorted-file order, a signature trailer -- which OpenNN's ImageDataset
    writes beside the class folders on its first pass at a given size. Feeding
    PyTorch from the same file isolates the input-pipeline asymmetry from the
    convolution throughput; it requires OpenNN to have opened the folder at
    this size once. Labels still come from ImageFolder, whose class and file
    order the cache shares.
    """

    def __init__(self, path: str, size: int):
        folder = Folders(path)
        self.targets = folder.targets
        self.shape = (size, size, 3)
        self.bytes = size * size * 3
        self.cache = os.path.join(path, ".cache", "images.bin")
        expected = len(self.targets) * self.bytes
        actual = os.path.getsize(self.cache) if os.path.exists(self.cache) else 0
        if actual < expected:
            raise SystemExit(f"{self.cache}: {actual} bytes, at least {expected} expected "
                             f"for {len(self.targets)} images at {size}x{size}")
        self.fd = None

    def __len__(self) -> int:
        return len(self.targets)

    def __getitem__(self, index: int):
        if self.fd is None:                        # one descriptor per worker process
            self.fd = os.open(self.cache, os.O_RDONLY)
        raw = os.pread(self.fd, self.bytes, index * self.bytes)
        pixels = torch.frombuffer(bytearray(raw), dtype=torch.uint8).view(self.shape)
        return pixels.permute(2, 0, 1).float().div_(255.0), self.targets[index]

def loader_for(path: str, batch: int, opts: dict, shuffle: bool) -> DataLoader:
    """Class folders, decoded per batch by worker processes -- or, with
    PT_INPUT=cache, OpenNN's pre-decoded cache read by the same workers."""
    if INPUT == "cache":
        dataset = CachedImages(path, opts["size"])
    else:
        dataset = Folders(path, transforms.Compose([
            transforms.Resize(opts["size"]),
            transforms.CenterCrop(opts["size"]),
            transforms.ToTensor(),
        ]))
    return DataLoader(dataset, batch_size=batch, shuffle=shuffle,
                      num_workers=WORKERS, pin_memory=True, drop_last=True,
                      persistent_workers=WORKERS > 0)

def autocast_ctx(opts: dict):
    if not opts["autocast"]:
        return contextlib.nullcontext()
    return torch.autocast(device_type=opts["device"], dtype=torch.bfloat16)

def compiled(fn, opts: dict):
    """torch.compile on CUDA, eager on CPU -- both measured, not assumed.

    On CPU, eager is PyTorch's fast path for these models, not a shortcut:
    inductor's CPU codegen loses on a small stack of GEMMs. Measured on this
    machine, dense CPU training, batch 4,096: eager 93,156 samples/s against
    compiled 83,722. The previous suite measured the same thing on different
    hardware (41,523 against 29,449), so it is the codegen and not the box.

    Compiling here anyway would hand OpenNN a win against a PyTorch nobody
    would ship, which is the mirror image of the eager-on-GPU mistake that
    made dense training read 1.29x when it was 1.06x.

    PT_COMPILE_MODE overrides either way, so the choice stays measurable.
    """
    mode = os.environ.get("PT_COMPILE_MODE", "default")
    if mode == "eager" or opts["device"] != "cuda":
        return fn, "eager"
    return torch.compile(fn, mode=None if mode == "default" else mode), f"compile:{mode}"

def batches_of(text: str) -> list[int]:
    return [int(part) for part in text.split(",") if part]

def sync(opts: dict) -> None:
    if opts["device"] == "cuda":
        torch.cuda.synchronize()

def to_device(images, opts: dict):
    if opts["device"] != "cuda":
        return images
    return images.to("cuda", non_blocking=True, memory_format=torch.channels_last)

def train_like(argv: list[str], mode: str) -> int:
    epochs = int(argv[4]) if len(argv) > 4 else 1
    batches = batches_of(argv[5] if len(argv) > 5 else "128")
    opts = parse_opts(argv, 6)

    print(f"baseline_rss_mib={resident_mib():.1f}")
    print(f"engine=pytorch\nmode={mode}\ndevice={opts['device']}")
    report_blas()
    warmup = 1 if mode == "train" else 0

    for batch in batches:
        loader = loader_for(argv[2], batch, opts, shuffle=True)
        print(f"samples={len(loader) * batch}")

        model = build(opts)
        if batch == batches[0]:
            print(f"parameters={sum(p.numel() for p in model.parameters())}", flush=True)
        loss_fn = torch.nn.CrossEntropyLoss()
        optimizer = torch.optim.Adam(model.parameters())
        ctx = autocast_ctx(opts)

        def step(images, labels):
            optimizer.zero_grad(set_to_none=True)
            with ctx:
                loss = loss_fn(model(images), labels)
            loss.backward()
            optimizer.step()

        step_fn, how = compiled(step, opts)

        def run_epoch():
            model.train()
            for images, labels in loader:
                step_fn(to_device(images, opts), labels.to(opts["device"], non_blocking=True))
            sync(opts)

        for _ in range(warmup):
            run_epoch()

        print(f"TIMED_START_UNIX={time.time():.3f}", flush=True)
        times = []
        for _ in range(epochs):
            mark = time.perf_counter()
            run_epoch()
            times.append(time.perf_counter() - mark)
        print(f"TIMED_END_UNIX={time.time():.3f}", flush=True)

        times.sort()
        median = times[len(times) // 2]

        print(f"batch_{batch}_samples_per_sec={int(len(loader) * batch / median)}"
              f" median_epoch_s={median:.6g}")
        print(f"batch_{batch}_epoch_times={','.join(f'{t:.6g}' for t in times)}")
        print(f"batch_{batch}_mode={how}", flush=True)

    print("RESULT=OK")
    return 0

def infer(argv: list[str]) -> int:
    reps = int(argv[3]) if len(argv) > 3 else 1
    batches = batches_of(argv[4] if len(argv) > 4 else "128")
    opts = parse_opts(argv, 5)

    print(f"baseline_rss_mib={resident_mib():.1f}")
    print(f"engine=pytorch\nmode=infer\ndevice={opts['device']}")
    report_blas()

    for batch in batches:
        loader = loader_for(argv[2], batch, opts, shuffle=False)
        samples = len(loader) * batch
        print(f"samples={samples}")

        model = build(opts).eval()
        weights_bf16 = opts["autocast"] and INFER_CAST == "weights"
        if weights_bf16:
            model = model.to(torch.bfloat16)
            opts = dict(opts, autocast=False)
        if batch == batches[0]:
            print(f"parameters={sum(p.numel() for p in model.parameters())}", flush=True)
        forward, _ = compiled(model, opts)

        # One batch, filled once and replayed, matching cnn.cpp: this times the
        # resident forward pass rather than the image decode, which the
        # training cell already accounts for.
        images = to_device(next(iter(loader))[0], opts)
        if weights_bf16:
            images = images.to(torch.bfloat16)

        def run_pass():
            with torch.no_grad(), autocast_ctx(opts):
                for _ in range(len(loader)):
                    forward(images)
            sync(opts)

        run_pass()

        print(f"TIMED_START_UNIX={time.time():.3f}", flush=True)
        times = []
        for _ in range(reps):
            mark = time.perf_counter()
            run_pass()
            times.append(time.perf_counter() - mark)
        print(f"TIMED_END_UNIX={time.time():.3f}", flush=True)

        print(f"batch_{batch}_pass_times={','.join(f'{t:.6g}' for t in times)}")
        times.sort()
        median = times[len(times) // 2]
        print(f"batch_{batch}_samples_per_sec={int(samples / median)}"
              f" median_pass_s={median:.6g}", flush=True)

    print("RESULT=OK")
    return 0

def capacity(argv: list[str]) -> int:
    batch = int(argv[3]) if len(argv) > 3 else 128
    opts = parse_opts(argv, 4)

    print(f"baseline_rss_mib={resident_mib():.1f}")
    print(f"engine=pytorch\nmode=capacity\ndevice={opts['device']}\nbatch={batch}")
    report_blas()

    try:
        loader = loader_for(argv[2], batch, opts, shuffle=False)
        images, labels = next(iter(loader))

        model = build(opts)
        print(f"parameters={sum(p.numel() for p in model.parameters())}", flush=True)
        loss_fn = torch.nn.CrossEntropyLoss()
        optimizer = torch.optim.Adam(model.parameters())

        with autocast_ctx(opts):
            loss = loss_fn(model(to_device(images, opts)),
                           labels.to(opts["device"], non_blocking=True))
        loss.backward()
        optimizer.step()
        sync(opts)
    except torch.cuda.OutOfMemoryError as error:
        print(f"fits=0\nreason={error}\nRESULT=OOM", flush=True)
        return 1
    except RuntimeError as error:
        if "out of memory" not in str(error).lower():
            raise
        print(f"fits=0\nreason={error}\nRESULT=OOM", flush=True)
        return 1

    print("fits=1\nRESULT=OK", flush=True)
    return 0

def main() -> int:
    argv = sys.argv
    mode = argv[1] if len(argv) > 1 else ""

    if mode in ("train", "quality") and len(argv) > 3:
        return train_like(argv, mode)
    if mode == "infer" and len(argv) > 2:
        return infer(argv)
    if mode == "capacity" and len(argv) > 2:
        return capacity(argv)

    print(__doc__, file=sys.stderr)
    return 2

if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/families/cnn.cpp
// The CNN family, defined once, driven four ways.
//
// PLAN.md. ResNet-50 v1.5: bottleneck blocks [3,4,6,3], widths
// [64,128,256,512], on the pinned ImageNet subset -- 1000 classes at 50
// images each, 224x224. All 1000 classes are kept so the head is the real
// 2048x1000; a ten-class subset would be a different network.
//
//   cnn train    <train_dir> <test_dir> [epochs] [batch,...] [size] [dev] [prec]
//   cnn infer    <test_dir>             [reps]   [batch,...] [size] [dev] [prec]
//   cnn capacity <train_dir>            [batch]              [size] [dev] [prec]
//   cnn quality  <train_dir> <test_dir> [epochs] [batch]      [size] [dev] [prec]
//
// Images are lazy-loaded per batch from class folders, which both engines do,
// so this measures convolution throughput *plus* input-pipeline efficiency.
// That is deliberate: at 50,000 x 224x224x3 the split cannot be resident, so
// pretending otherwise would measure a workload nobody runs.

#include <algorithm>
#include <numeric>
#include <chrono>
#include <iomanip>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#ifdef __linux__
#include <unistd.h>
#endif

#include "opennn/core/configuration.h"
#include "opennn/core/tensor_operations.h"
#include "opennn/core/random_utilities.h"
#include "opennn/core/device_backend.h"
#include "opennn/core/tensor_types.h"
#include "opennn/dataset/image_dataset.h"
#include "opennn/neural_network/forward_propagation.h"
#include "opennn/models/models.h"
#include "opennn/training_strategy/adaptive_moment_estimation.h"
#include "opennn/training_strategy/training_strategy.h"

using namespace opennn;
using clock_type = chrono::steady_clock;

namespace
{

double resident_mb()
{
#ifdef __linux__
    // statm reports pages; the second field is resident.
    ifstream statm("/proc/self/statm");
    long total = 0, resident = 0;
    if (statm >> total >> resident)
        return double(resident) * double(sysconf(_SC_PAGESIZE)) / (1024.0 * 1024.0);
#endif
    return 0.0;
}

constexpr Index SEED = 42;

struct Options
{
    Index image_size = 224;
    Device device = Device::CUDA;
    Type precision = Type::FP32;
};

unique_ptr<ResNet> build(ImageDataset& dataset)
{
    set_seed(SEED);

    return make_unique<ResNet>(dataset.get_shape("Input"),
                               vector<Index>{3, 4, 6, 3},
                               Shape{64, 128, 256, 512},
                               dataset.get_shape("Target"),
                               true);
}

unique_ptr<ImageDataset> open_dataset(const string& path, const Options& options)
{
    return options.image_size > 0
        ? make_unique<ImageDataset>(path, Shape{options.image_size, options.image_size, 3})
        : make_unique<ImageDataset>(path);
}

vector<Index> parse_batches(const string& text)
{
    vector<Index> batches;
    stringstream stream(text);

    for (string item; getline(stream, item, ',');)
        if (!item.empty())
            batches.push_back(stoll(item));

    return batches;
}

Options parse_options(int argc, char* argv[], int first)
{
    Options options;

    if (argc > first)     options.image_size = stoll(argv[first]);
    if (argc > first + 1) options.device = string(argv[first + 1]) == "cpu" ? Device::CPU : Device::CUDA;
    if (argc > first + 2) options.precision = string(argv[first + 2]) == "bf16" ? Type::BF16 : Type::FP32;

    return options;
}

AdaptiveMomentEstimation* configure(TrainingStrategy& strategy, Index batch)
{
    strategy.set_loss("CrossEntropy");
    strategy.get_loss()->set_regularization("NoRegularization");
    strategy.set_optimization_algorithm("AdaptiveMomentEstimation");

    auto* adam = dynamic_cast<AdaptiveMomentEstimation*>(strategy.get_optimization_algorithm());
    adam->set_batch_size(batch);
    adam->set_display_period(1000000);
    adam->set_gradient_clip_norm(0.0f);

    return adam;
}

int usage()
{
    cerr << "usage: cnn train    <train_dir> <test_dir> [epochs] [batch,...] [size] [dev] [prec]\n"
            "       cnn infer    <test_dir>             [reps]   [batch,...] [size] [dev] [prec]\n"
            "       cnn capacity <train_dir>            [batch]              [size] [dev] [prec]\n";
    return 2;
}

}   // namespace

int main(int argc, char* argv[])
{
    // Each engine at its best, as PROTOCOL.md requires. The library defaults to
    // Eigen so a plain build behaves like a plain build; a build that has the
    // MKL kernels is told to use them here rather than inheriting them.
    Configuration::instance().set_blas(Blas::Mkl);
    cout << "blas=" << (blas_mkl_available() ? "mkl" : "eigen") << "\n";

    const string mode = argc > 1 ? argv[1] : "";

    if (mode == "train" || mode == "quality")
    {
        if (argc < 4) return usage();

        const Index epochs = argc > 4 ? stoll(argv[4]) : 1;
        const vector<Index> batches = parse_batches(argc > 5 ? argv[5] : "128");
        const Options options = parse_options(argc, argv, 6);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=" << mode
             << "\ndevice=" << (options.device == Device::CPU ? "cpu" : "cuda") << "\n";

        auto dataset = open_dataset(argv[2], options);
        dataset->set_sample_roles("Training");

        const Index all_samples = dataset->get_samples_number();
        const bool timing = mode == "train";
        const Index warmup = timing ? 2 : 0;

        cout << "samples=" << all_samples << "\n";

        for (const Index batch : batches)
        {
            // Whole batches only, matching the PyTorch driver's
            // range(0, n - batch + 1, batch). A tail batch is real work the
            // throughput figure does not count, and the library keeps a second
            // set of activation contexts for it.
            dataset->set_sample_roles("Training");
            for (Index sample = (all_samples / batch) * batch; sample < all_samples; ++sample)
                dataset->set_sample_role(sample, SampleRole::None);

            auto network = build(*dataset);
            if (batch == batches.front())
                cout << "parameters=" << network->get_parameters_number() << "\n" << flush;

            const bool graph = options.device == Device::CUDA
                               && getenv("OPENNN_NO_CUDA_GRAPH") == nullptr;

            TrainingStrategy strategy(network.get(), dataset.get());
            auto* adam = configure(strategy, batch);
            adam->set_cuda_graph(graph);
            adam->set_maximum_epochs(warmup + epochs);

            const auto unix_now = []
            {
                return chrono::duration<double>(
                    chrono::system_clock::now().time_since_epoch()).count();
            };

            vector<double> epoch_seconds;
            auto previous_mark = clock_type::now();

            adam->post_epoch_callback = [&](Index epoch, float, float, NeuralNetwork*)
            {
                const auto now = clock_type::now();
                const double elapsed = chrono::duration<double>(now - previous_mark).count();
                previous_mark = now;

                if (epoch == warmup - 1)
                    cout << "TIMED_START_UNIX=" << fixed << setprecision(3)
                         << unix_now() << "\n" << defaultfloat;
                else if (epoch >= warmup)
                    epoch_seconds.push_back(elapsed);

                if (epoch == warmup + epochs - 1)
                    cout << "TIMED_END_UNIX=" << fixed << setprecision(3)
                         << unix_now() << "\n" << defaultfloat;
            };

            strategy.train();

            if (Index(epoch_seconds.size()) != epochs)
            {
                cerr << "epoch timing marks missing\n";
                return 1;
            }

            sort(epoch_seconds.begin(), epoch_seconds.end());
            const double median_epoch_s = epoch_seconds[epoch_seconds.size() / 2];
            const Index samples_per_epoch = (all_samples / batch) * batch;

            cout << "batch_" << batch << "_samples_per_sec="
                 << long(double(samples_per_epoch) / median_epoch_s)
                 << " median_epoch_s=" << median_epoch_s << "\n"
                 << "batch_" << batch << "_epoch_times=";
            for (size_t i = 0; i < epoch_seconds.size(); ++i)
                cout << (i ? "," : "") << epoch_seconds[i];
            cout << "\n"
                 << "batch_" << batch << "_cuda_graph="
                 << (!graph ? "off"
                     : adam->get_cuda_graph_capture_failed() ? "failed" : "captured")
                 << "\n" << flush;
        }
    }
    else if (mode == "infer")
    {
        if (argc < 3) return usage();

        const Index reps = argc > 3 ? stoll(argv[3]) : 1;
        const vector<Index> batches = parse_batches(argc > 4 ? argv[4] : "128");
        const Options options = parse_options(argc, argv, 5);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=infer\ndevice="
             << (options.device == Device::CPU ? "cpu" : "cuda") << "\n";

        auto dataset = open_dataset(argv[2], options);
        dataset->set_sample_roles("Testing");

        const Index samples = dataset->get_samples_number();
        auto network = build(*dataset);

        cout << "samples=" << samples
             << " parameters=" << network->get_parameters_number() << "\n" << flush;

        for (const Index batch : batches)
        {
            const Index processed = (samples / batch) * batch;
            ForwardPropagation forward_propagation(batch, network.get(),
                                                  ForwardPropagationMode::Inference);
            forward_propagation.set_cuda_graph(options.device == Device::CUDA);

            // One batch is filled once and replayed: this measures the
            // resident forward pass, the same thing the PyTorch driver times
            // after its own warmup, rather than the image decode that
            // training already accounts for.
            vector<Index> indices(size_t(batch), Index(0));
            iota(indices.begin(), indices.end(), Index(0));

            Batch data(batch, dataset.get(), network->get_config());
            data.fill(indices, dataset->get_feature_selection(), FillMode::Inference);

            const vector<TensorView>& inputs = data.get_inputs();

            const auto run_pass = [&]
            {
                for (Index i = 0; i + batch <= samples; i += batch)
                    network->calculate_outputs_resident(inputs, forward_propagation, false);

                // The clock stops when the work is done, not when it is
                // queued, matching torch.cuda.synchronize() in the PyTorch
                // driver. Without it this cell reported passes of 0.0019 s,
                // 9.69 s and 25.3 s for identical work: the first pass timed
                // the launches and the ones after it paid for the backlog.
                if (options.device == Device::CUDA)
                    device::synchronize(device::get_compute_stream());
            };

            network->calculate_outputs_resident(inputs, forward_propagation, true);
            run_pass();

            const auto unix_now = []
            {
                return chrono::duration<double>(
                    chrono::system_clock::now().time_since_epoch()).count();
            };

            cout << "TIMED_START_UNIX=" << fixed << setprecision(3)
                 << unix_now() << "\n" << defaultfloat << flush;

            vector<double> times;
            for (Index r = 0; r < reps; ++r)
            {
                const auto t0 = clock_type::now();
                run_pass();
                times.push_back(chrono::duration<double>(clock_type::now() - t0).count());
            }

            cout << "TIMED_END_UNIX=" << fixed << setprecision(3)
                 << unix_now() << "\n" << defaultfloat << flush;

            cout << "batch_" << batch << "_pass_times=";
            for (size_t i = 0; i < times.size(); ++i) cout << (i ? "," : "") << times[i];
            cout << "\n";

            sort(times.begin(), times.end());
            const double median_pass_s = times[times.size() / 2];

            cout << "batch_" << batch << "_samples_per_sec="
                 << long(double(processed) / median_pass_s)
                 << " median_pass_s=" << median_pass_s << "\n" << flush;
        }
    }
    else if (mode == "capacity")
    {
        if (argc < 3) return usage();

        const Index batch = argc > 3 ? stoll(argv[3]) : 128;
        const Options options = parse_options(argc, argv, 4);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=capacity\ndevice="
             << (options.device == Device::CPU ? "cpu" : "cuda")
             << "\nbatch=" << batch << "\n";

        // One attempt, then exit: an out-of-memory fault leaves the CUDA
        // context unusable, so a second attempt here would measure the wreck
        // of the first.
        try
        {
            auto dataset = open_dataset(argv[2], options);
            dataset->set_sample_roles("Training");

            auto network = build(*dataset);
            cout << "parameters=" << network->get_parameters_number() << "\n" << flush;

            TrainingStrategy strategy(network.get(), dataset.get());
            configure(strategy, batch)->set_maximum_epochs(1);
            strategy.train();
        }
        catch (const exception& error)
        {
            cout << "fits=0\nreason=" << error.what() << "\nRESULT=OOM\n" << flush;
            return 1;
        }

        cout << "fits=1\nRESULT=OK\n" << flush;
        return 0;
    }
    else
    {
        return usage();
    }

    cout << "RESULT=OK\n";

    return 0;
}
